polynote-spark/src/main/scala/polynote/kernel/remote/RemoteSparkKernel.scala

```diff
@@ -0,0 +1,337 @@
+package polynote.kernel.remote
+
+import java.net.{InetSocketAddress, URI, URISyntaxException}
+import java.nio.ByteBuffer
+import java.nio.channels._
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+import java.util.concurrent.{ConcurrentHashMap, ExecutorService, Executors, LinkedBlockingQueue}
+
+import cats.effect.concurrent.{Deferred, Semaphore}
+import cats.effect.{ContextShift, Fiber, IO, Timer}
+import cats.syntax.apply._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.traverse._
+import cats.instances.list._
+import fs2.{Chunk, Stream}
+import fs2.concurrent.Queue
+import org.log4s.getLogger
+import polynote.config.PolynoteConfig
+import polynote.kernel._
+import polynote.kernel.util.{Publish, ReadySignal}
+import polynote.messages.{ByteVector32, CellID, CellResult, HandleType, Lazy, Notebook, NotebookUpdate, ShortString, Streaming, Updating}
+import polynote.runtime._
+
+import scala.concurrent.duration.{Duration, MINUTES}
+
+class RemoteSparkKernel(
+  val statusUpdates: Publish[IO, KernelStatusUpdate],
+  getNotebook: () => IO[Notebook],
+  config: PolynoteConfig,
+  transport: TransportServer)(implicit
+  contextShift: ContextShift[IO]
+) extends KernelAPI[IO] {
+
+  private val remoteAddr = Deferred.unsafe[IO, String]
+  private val mainFiber = Deferred.unsafe[IO, Fiber[IO, Unit]]
+  private val shutdownSignal = ReadySignal()
+
+  private val requestId = new AtomicInteger(0)
+  private val requests = new ConcurrentHashMap[Int, Deferred[IO, RemoteResponse]]()
+  private val streamRequests = new ConcurrentHashMap[Int, Deferred[IO, Queue[IO, Option[Result]]]]
+
+  private val logger = getLogger
+
+  /**
+    * Send the given message, expecting the given response, and return an IO representing the response being received.
+    * The outer IO of the return value sends the request, and the inner layer awaits the response.
+    */
+  private def send1[Rep <: RemoteResponse](req: RemoteRequest) = for {
```

Comment:
Would be useful to have the type signature here since its referenced in the comment.

Reply:
Heh, I didn't even notice it wasn't there because IntelliJ shows it anyway 😛 Will fix